### ISSUE TITLE
major differences between translation time and training time 

### ISSUE THREAD

**Issue by yftah89** [EdinburghNLP/nematus#27]
When I use translate.py the results are quite weird, if I train my model for low number of iterations I end up with blank translations(each sentence from the test is translated to eos right away).
If I train my model for high number of iterations I end up with sentences which contains single word with repetitions.

The validation examples during training time looks OK (--sampleFreq param).

Do you have any thoughts?

**Comment by bhaddow** [comment]
Hi

(Note that Nematus has a support list 
https://groups.google.com/d/forum/nematus-support)

When translation goes badly wrong like this, it's usually a data 
problem. Is your data sentence aligned, is it in the correct language? 
How large is your training data, and where does it come from?

Does perplexity on the development set decrease (look at the training log)?

Post your command line, so we can see what settings you are using,

cheers - Barry

**Comment by yftah89** [comment]
Hi Barry, thanks for the quick response,

When translation goes badly wrong like this, it's usually a data
problem. Is your data sentence aligned, is it in the correct language? - _yes for both questions_
How large is your training data, and where does it come from? - _about 50K sentences, not taken from translation dataset_ 

Does perplexity on the development set decrease (look at the training log)? _yes it does, the weird thing is that during training time I get some samples from my validation set in the following way :
source: [the source sentence]
truth: [the real translation]
sample: [the model's translation] 
and it seems to be allright (I don't refer to actual translation's accuracy I rather refer to simple sanity check - different words and not just empty line or one repetitive word like translation time).
So I don't believe I have over\under fitting problem._ 

_the training command is the following:_ 

THEANO_FLAGS=device=gpu,floatX=float32  python nmt.py --datasets "examples-train.text" "labels-train.text" --dictionaries "examples-train.text.json" "labels-train.text.json" --model check.model --dim_word 100 --dim 64  --valid_datasets "examples-val.text" "labels-val.text" 

_the translating command is the following:_ 

 THEANO_FLAGS=device=gpu,floatX=float32  python translate.py --input examples-test.text --output answers.txt --model check.model.npz

I know that --dim might seems to be small but I really don't think its underfitting problem, moreover, when I try to translate my validation set (the same one I get OK samples on in training time) I'm still getting the same weird results (all sentences are eos or on repetitive word)

**Comment by bhaddow** [comment]
Hi

What exactly is your training data, if it is not translation?

Have you tried decoding with the -n (length normalisation) flag?

What about an n-best list? Do you see sensible translations anywhere in 
the list?

cheers - Barry

**Comment by yftah89** [comment]
What exactly is your training data, if it is not translation?
_I'm trying to translate sentences from English to English (kind of paraphrasing)_

Have you tried decoding with the -n (length normalisation) flag?
What about an n-best list? Do you see sensible translations anywhere in
the list?
_I just tried both and unfortunately neither helps._  

What really bugs me is that in training time the translated samples looks great, what is the difference between whats going on in the training time and what going on in translation time? 
Is the samples which are given at training time are not translated in the exact same way like the translations in translation time? if not then what is the difference? 

Thanks in advance for all of your help

**Comment by bhaddow** [comment]
Hi

Sampling and translating use the same code, but sampling is stochastic,
whereas translating uses beam search to find the top scoring outputs.
That's why I was asking about an n-best list, maybe there are good
translations with worse model scores.

It does sound like your model is under-trained. Perhaps the data is not
sufficient, perhaps it is noisy, perhaps it is not the right model for this
task. When a translation model is under-trained it just latches onto words
or phrases from the target language and repeats them. Have a look at the
perplexities - have they converged? If yes, and the model is still
under-trained, then you probably need more data or a different architecture

cheers - Barry

**Comment by yftah89** [comment]
But if you are right and my model is under trained or my perplexities are not converged, why the samples that the model provides me in training time are OK? wouldn't bad model provide bad samples? 
If I could get the same results I get in samples for my whole test set I will be very happy

**Comment by bhaddow** [comment]
Hi

To be honest, I never look at the samples. How do the perplexities look?

cheers - Barry

**Comment by yftah89** [comment]
I'm not sure that perplexity is the right way to measure my performance, the loss that nematus provides is around 20, I'm not sure if this considered to be good or not.

Can I use some interface for stochastic inference in translation time or I'm bound to write my on code in order to do so? 

Again, thank for your help and efforts

**Comment by bhaddow** [comment]
Hi

It's not the absolute loss that matters, it's the convergence I was asking
about. Nematus optimises cross-entropy by default, which is closely related
to perplexity, so if it is not the right way to measure your performance,
then you have a problem.

You can could write some code to call the gen_sample() method in nmt.py, to
give stochastic samples. However this is not tackling the real problem -
beam search should be better than stochastic sampling,

cheers - Barry

**Comment by yftah89** [comment]
I think I have found the cause for the major differences between the samples in training time and the results in translation time but I want to be sure, can you please shed some light on this comments in the code 
#FIXME: sampling is currently performed according to the last model only
line 724 in nmt.py

**Comment by rsennrich** [comment]
The gen_sample function supports ensembles for beam search, but for ancestral sampling, doesn't support ensembles and uses only its last component if passed an ensemble. Unless you're using ensembles at test time, this difference shouldn't affect you.

**Comment by yftah89** [comment]
I see, so I still got no idea why I get bad results in translation time and good ones in the samples that the train time provides, the only difference I see so far is that the samples are stochastic and the translation use beam search but to the best of my knowledge beam search should be better, do you have any idea of some additional differences between the samples in training time and the translation process? 

If not, can you give me some explanation\references about stochastic vs beam search methods and if stochastic can be better than beam search?
